tests/v1/worker/test_prefix_caching_worker.py

```diff
@@ -0,0 +1,504 @@
+from dataclasses import fields
+from typing import Any, Optional
+
+import pytest
+import torch
+#from torch import nn
+from scheduling_utils import create_request_for_scheduler_test, random_prompt
+from spyre_util import ModelInfo, patch_environment
+from vllm import EngineArgs
+from vllm.config import VllmConfig
+from vllm.forward_context import get_forward_context
+from vllm.v1.core.sched.output import (CachedRequestData, NewRequestData,
+                                       SchedulerOutput)
+from vllm.v1.outputs import ModelRunnerOutput, SamplerOutput
+from vllm.v1.request import Request
+from vllm.v1.sample.metadata import SamplingMetadata
+from vllm.v1.sample.sampler import Sampler
+
+import vllm_spyre.envs as envs_spyre
+from vllm_spyre.model_executor.model_loader.spyre import SpyreAttentionMetadata
+from vllm_spyre.platform import SpyrePlatform
+from vllm_spyre.v1.worker.spyre_model_runner import ChunkedPrefillModelRunner
+
+
+class MockContinuousBatchingFmsModel:
+
+    def set_past_key_value_states(self, num_blocks) -> None:
+        pass
+
+
+class MockSpyreCausalLM:
+
+    def __init__(
+        self,
+        vllm_config: VllmConfig,
+    ) -> None:
+
+        try:
+            ## Temporary backwards compatibility for 0.10.2
+            from vllm.model_executor.layers.sampler import get_sampler
+            self.sampler = get_sampler()
+        except (ImportError, ModuleNotFoundError):
+            self.sampler = Sampler()
+
+        # boolean tensor of length batch size with indices:
+        # True for unfinished sequences and
+        # False for finished or padded sequences
+        self.indices = None
+
+        # number of right pads (relevant for continuous batching only)
+        self.n_pads_right = 0
+
+        self._mask_dtype = torch.float16 if \
+            envs_spyre.VLLM_SPYRE_DYNAMO_BACKEND == "sendnn" \
+            else torch.float32
+
+        self.model = MockContinuousBatchingFmsModel()
+
+        self.vocab_size = vllm_config.model_config.get_vocab_size()
+
+        self.last_input_ids: Optional[torch.Tensor] = None
+        self.last_positions: Optional[torch.Tensor] = None
+        self.last_masks: Optional[torch.Tensor] = None
+        self.last_is_prompt: Optional[bool] = None
+        self.last_attn_metadata: Optional[SpyreAttentionMetadata] = None
+
+    def __call__(self, *args, **kwargs):
+        return self.forward(*args, **kwargs)
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        masks: torch.Tensor,
+        is_prompt: bool,
+    ) -> torch.Tensor:
+
+        self.last_input_ids = input_ids
+        self.last_positions = positions
+        self.last_masks = masks
+        self.last_is_prompt = is_prompt
+
+        forward_context = get_forward_context()
+
+        assert isinstance(forward_context.attn_metadata,
+                          SpyreAttentionMetadata)
+        self.last_attn_metadata = forward_context.attn_metadata
```

Comment:
If `get_forward_context` is all we need to call at each step to get the step's attention metadata, then I think we should be able to do that in the existing scheduler step tests. The tests are at the engine level, but turn off multiprocessing so that the model runner and model are in the same process so we should be able to fetch the real forward context and save a copy of the attn_metadata at each step